cpp/src/slice2cs/Gen.cpp

```diff
@@ -2580,22 +2589,27 @@ Slice::Gen::HelperVisitor::visitSequence(const SequencePtr& p)
 {
     string name = p->name();
     string scope = getNamespace(p);
-    string seqS = typeToString(p, scope, p->hasMetaDataWithPrefix("cs:serializable:"));
-    TypePtr type = p->type();
+    string seqS = typeToString(p, scope);
+    string seqReadOnly = typeToString(p, scope, false, true);
 
     _out << sp;
     emitGeneratedCodeAttribute();
     _out << nl << "public static class " << name << "Helper";
     _out << sb;
 
     _out << sp;
-    _out << nl << "public static void Write(this Ice.OutputStream ostr, " << seqS << " sequence) => ";
+    _out << nl << "public static void Write(this Ice.OutputStream ostr, " << seqReadOnly << " sequence) => ";
     _out.inc();
     _out << nl << sequenceMarshalCode(p, scope, "sequence", "ostr") << ";";
     _out.dec();
 
     _out << sp;
-    _out << nl << "public static readonly Ice.OutputStreamWriter<" << seqS << "> IceWriter = Write;";
+     _out << nl << "public static readonly Ice.OutputStreamWriter<" << seqReadOnly << "> IceWriter = Write;";
+    // IceNestedWriter is used only when marshaling nested sequences, and we need them in the correct type because
+    // a T[] (where T : fixed size numeric type) is only implicitly converted to a ReadOnlyMemory<T>, there is no is-a
+    // relationship.
+    _out << nl << "public static readonly Ice.OutputStreamWriter<" << seqS
+        << "> IceNestedWriter = (ostr, sequence) => Write(ostr, sequence);";
```

Comment:
Should we generate this conditionally, seems it is only required when sequence element is  fixed size numeric type?

Reply:
I am replacing these nested writers by a few "array" writers on OutputStream.